src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/BatchEmitter.java

```diff
@@ -163,5 +166,17 @@ private SelfDescribingJson getFinalPost(List<TrackerPayload> buffer) {
     @Override
     public void close() {
         flushBuffer();
+        if (executor != null) {
+            executor.shutdown();
+            try {
+                while (!executor.awaitTermination(closeTimeout, TimeUnit.SECONDS)) {
+                    LOGGER.debug("Executor doesn't shut down yet");
+                }
+                LOGGER.debug("Executor service shut down");
+            } catch (InterruptedException e) {
+                LOGGER.error("Encounter an interrupted exception", e);
+            }
```

Comment:
I don't think the while is good idea since you might end up with a loop running forever if a task doesn't terminate.
I think if the first `awaittermination` is not successful we'll have to end the tasks ourselves, following the guidelines in the link above:

```java
if (executor != null) {
   executor.shutdown();
   try {
     if (!executor.awaitTermination(closeTimeout, TimeUnit.SECONDS)) {
       executor.shutdownNow();
       if (!executor.awaitTermination(closeTimeout, TimeUnit.SECONDS))
           LOGGER.warn("Executor did not terminate");
     }
   } catch (InterruptedException ie) {
     executor.shutdownNow();
     Thread.currentThread().interrupt();
   }
}
```